### ISSUE TITLE
Runtime segfault if a function returns an array of any kind.

### ISSUE THREAD

**Issue by p-pap** [lfortran/lfortran#2946]
The following works as expected:
```
program test
implicit none
print *,foo()

contains

function foo() result(res)
character(len=:), allocatable :: res
res="foobar"
end function foo

end program test
```
However, if `foo` returns an allocatable array of any kind (integers, reals, or even characters) it compiles but causes a segmentation fault at runtime, e.g.:
```
function foo() result(res)
integer, dimension(:), allocatable :: res
res=[1, 2, 3, 4]
end function foo
```
or
```
function foo() result(res)
real, dimension(:), allocatable :: res
res=real([1, 2, 3, 4])
end function foo
```
Any result that has the `dimension` attribute fails.
The problem is probably related to [#2370](https://github.com/lfortran/lfortran/issues/2370#issue-1882585982) except it doesn't happen in a module, and it doesn't issue an error, just segfaults at runtime.

**Comment by certik** [comment]
You have to use `--realloc-lhs` like this:
```console
$ lfortran --realloc-lhs a.f90 
1 2 3 4 
```
or allocate the array to the right size explicitly:
```fortran
function foo() result(res)
integer, dimension(:), allocatable :: res
allocate(res(4))
res=[1, 2, 3, 4]
end function foo
```
Then it works.

In the default Debug mode we need to raise a nice runtime error if the array is not allocated, not segfault like this. I'll change the title of this bug to reflect this.

**Comment by p-pap** [comment]
As you know, explicit allocation of arrays is not mandatory, as long the standard is F2003 or newer. In fact, it's common practice nowadays to use automatic allocation, unless there is a reason not to do so (I rarely have to do that, but it happens.)
I assume in Release mode the code above should be valid?

**Comment by certik** [comment]
Yes, we are keeping the door open to make `--realloc-lhs` the default, that would not break anything. The reason we put it behind an option for now is performance, so far I don't know how to implement it without a runtime check. Once we can compile most codes, we will focus on performance and make a decision here. 

I do not want to compromise performance, so if it turns out that there is a performance implication, then I think it should not be on by default. Users should explicitly enable it, if they want it and understand the performance implications.

**Comment by p-pap** [comment]
> I do not want to compromise performance, so if it turns out that there is a performance implication, then I think it should not be on by default. Users should explicitly enable it, if they want it and understand the performance implications.

Makes perfect sense, thank you for the clarification.
